website/addons/base/__init__.py

```diff
@@ -350,6 +351,46 @@ def deep_url(self):
     def revision(self):
         return getattr(self, '_revision', None)
 
+    def add_tag(self, tag, auth, node, file_name, save=True):
+        if tag not in self.tags:
+            new_tag = Tag.load(tag)
+            if not new_tag:
+                new_tag = Tag(_id=tag)
+            new_tag.save()
+            self.tags.append(new_tag)
+            node.add_log(
+                action=NodeLog.FILETAG_ADDED,
+                params={
+                    'node': node._primary_key,
+                    'fileName': file_name,
+                    'tag': tag,
+                    'url': self.guid_url
+                },
+                auth=auth,
+                save=False,
+            )
+            if save:
+                self.save()
+                node.save()
```

Comment:
If we do keep the `save` flag, I don't believe the `node` should ever *not* be saved - the `save` flag on the methods of `Node` apply only to that object, not to other objects that may have been modified by the method.

One reason for this is that you may not have a reference to the external object in the context where the method is called.